src/Umbraco.Infrastructure/Persistence/Repositories/Implement/SimilarNodeName.cs

```diff
@@ -1,118 +1,249 @@
-                        return xname[i] < yname[i] ? -1 : +1;
-                    i++;
+                    current++;
+                }
+                else if (item.Number > current)
+                {
+                    // do nothing - we found our number!
+                    // eg. when suffixes are 1 & 3, then this method is required to generate 2
+                    break;
                 }
-                return 0;
             }
+
+            return current;
         }
 
-        // gets a unique name
-        public static string GetUniqueName(IEnumerable<SimilarNodeName> names, int nodeId, string nodeName)
+        internal class StructuredName
         {
-            var uniqueNumber = 1;
-            var uniqueing = false;
-            foreach (var name in names.OrderBy(x => x, new Comparer()))
+            private const string Suffixed_Pattern = @"(.*)\s\((\d+)\)$";
+            internal string Text { get; private set; }
+            internal int Number { get; private set; }
+            public string FullName
+            {
+                get
+                {
+                    string text = string.IsNullOrWhiteSpace(Text) ? Text.Trim() : Text;
+
+                    return Number > 0 ? $"{text} ({Number})" : text;
+                }
+            }
+
+            internal StructuredName(string name)
             {
-                // ignore self
-                if (nodeId != 0 && name.Id == nodeId) continue;
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    Text = SPACE_CHARACTER;
+                    Number = 0;
 
-                if (uniqueing)
+                    return;
+                }
+
+                var rg = new Regex(Suffixed_Pattern);
+                var matches = rg.Matches(name);
+                if (matches.Count > 0)
                 {
-                    if (name.NumPos > 0 && name.Name.InvariantStartsWith(nodeName) && name.NumVal == uniqueNumber)
-                        uniqueNumber++;
-                    else
-                        break;
+                    var match = matches[0];
+                    Text = match.Groups[1].Value;
+                    int number;
+                    Number = int.TryParse(match.Groups[2].Value, out number) ? number : 0;
+
+                    return;
                 }
-                else if (name.Name.InvariantEquals(nodeName))
+                else
                 {
-                    uniqueing = true;
+                    Text = name;
+                    Number = 0;
                 }
             }
 
-            return uniqueing || string.IsNullOrWhiteSpace(nodeName)
-                ? string.Concat(nodeName, " (", uniqueNumber.ToString(), ")")
-                : nodeName;
+            internal bool HasSuffix()
+            {
+                return Number > 0;
+            }
+
+            internal void SetSuffix(int number)
+            {
+                Number = number;
```

Comment:
We should check that number > 0. Or at least not negative :)
```suggestion
            {
                if(number <= 0) {
                    throw new ArgumentException("Suffix must be a positive number", nameof(number));
                }
                Number = number;
```

Reply:
This is yet another edge case which probably no one has considered. If a user decides to name their node "Test (-1)" then we should allow it. We should assume that "Test (-1)" is a complete string which the user wants to use. 

The suffix detecting regex will just need to be tweaked to only match on positive numbers. No match will default to using the entire string as the text. If you're ok with this, I will make the change.